worker/remoterelations/remoterelations.go

```diff
@@ -91,24 +93,312 @@ type Worker struct {
+	appWorker, ok := w.applicationWorkers[name]
+	if ok {
+		// We can ignore any error as the worker is still
+		// tracked by the catacomb.
+		worker.Stop(appWorker)
+		delete(w.applicationWorkers, name)
+	}
+}
+
+// remoteApplicationWorker listens for changes to relations
+// on a remote application and handles the publication of
+// those changes to the remote model.
+type remoteApplicationWorker struct {
+	catacomb         catacomb.Catacomb
+	relationsWatcher watcher.StringsWatcher
+	facade           RemoteApplicationsFacade
+}
+
+type relation struct {
+	params.RemoteRelationChange
+	ruw *relationUnitsWatcher
+}
+
+func newRemoteApplicationWorker(
+	relationsWatcher watcher.StringsWatcher,
+	facade RemoteApplicationsFacade,
+
+) (worker.Worker, error) {
+	w := &remoteApplicationWorker{
+		relationsWatcher: relationsWatcher,
+		facade:           facade,
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+	if err == nil {
+		// We want the relations watcher bound to the lifecycle of this worker.
+		if err := w.catacomb.Add(relationsWatcher); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return w, err
+}
+
+// Kill is defined on worker.Worker
+func (w *remoteApplicationWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is defined on worker.Worker
+func (w *remoteApplicationWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *remoteApplicationWorker) loop() error {
+	relations := make(map[string]*relation)
 	for {
-		// TODO(wallyworld)
 		select {
 		case <-w.catacomb.Dying():
+			// We can ignore any error as the worker is still
+			// tracked by the catacomb.
 			return w.catacomb.ErrDying()
+		case change, ok := <-w.relationsWatcher.Changes():
+			logger.Debugf("relations changed: %#v, %v", change, ok)
+			if !ok {
+				// We are dying.
+				continue
+			}
+			results, err := w.facade.RemoteRelations(change)
+			if err != nil {
+				return errors.Annotate(err, "querying relations")
+			}
+			for i, result := range results {
+				key := change[i]
+				if err := w.relationChanged(key, result, relations); err != nil {
+					return errors.Annotatef(err, "handling change for relation %q", key)
+				}
+			}
+		}
+	}
+}
+
+func (w *remoteApplicationWorker) killRelationUnitWatcher(key string, relations map[string]*relation) {
+	relation, ok := relations[key]
+	if ok {
+		// We can ignore any error as the worker is still
+		// tracked by the catacomb.
+		worker.Stop(relation.ruw)
+		delete(relations, key)
+	}
+}
+
+func (w *remoteApplicationWorker) relationChanged(
+	key string, result params.RemoteRelationResult, relations map[string]*relation,
+) error {
+	logger.Debugf("relation %q changed: %+v", key, result.Result)
+	if result.Error != nil {
+		if params.IsCodeNotFound(result.Error) {
+			// TODO(wallyworld) - once a relation dies, wait for
+			// it to be unregistered from remote side and then use
+			// cleanup to remove.
+			w.killRelationUnitWatcher(key, relations)
```

Comment:
The firewaller worker uses this pattern, as does the catacomb itself. And several other places.It seems like the way things are done. I'm not sure what we'd do with the error?